x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java

```diff
@@ -0,0 +1,526 @@
+            return;
+        }
+
+        EncodedVersion encoding = encodeVersion(versionString);
+        BytesRef encodedVersion = encoding.bytesRef;
+        if (fieldType.indexOptions() != IndexOptions.NONE || fieldType.stored()) {
+            Field field = new Field(fieldType().name(), encodedVersion, fieldType);
+            context.doc().add(field);
+            // encode the first 16 bytes as points for efficient range query
+            byte[] first16bytes = Arrays.copyOfRange(encodedVersion.bytes, encodedVersion.offset, 16);
+            context.doc().add(new BinaryPoint(fieldType().name(), first16bytes));
+        }
+        context.doc().add(new SortedSetDocValuesField(fieldType().name(), encodedVersion));
+
+        // add additional information extracted from version string
+        context.doc().add(new Field(prereleaseSubField.name(), encoding.isPreRelease ? "T" : "F", BooleanFieldMapper.Defaults.FIELD_TYPE));
+        context.doc().add(new SortedNumericDocValuesField(prereleaseSubField.name(), encoding.isPreRelease ? 1 : 0));
+
+        addVersionPartSubfield(context, majorVersionSubField.name(), encoding.major);
+        addVersionPartSubfield(context, minorVersionSubField.name(), encoding.minor);
+        addVersionPartSubfield(context, patchVersionSubField.name(), encoding.patch);
+    }
+
+    private void addVersionPartSubfield(ParseContext context, String fieldName, Integer versionPart) {
+        if (versionPart != null) {
+            context.doc().addAll(NumberType.INTEGER.createFields(fieldName, versionPart, true, true, false));
+        }
+    }
+
+    @Override
+    protected void mergeOptions(FieldMapper other, List<String> conflicts) {
+        VersionStringFieldMapper mergeWith = (VersionStringFieldMapper) other;
+        this.nullValue = mergeWith.nullValue;
+    }
+
+    @Override
+    protected void doXContentBody(XContentBuilder builder, boolean includeDefaults, Params params) throws IOException {
+        super.doXContentBody(builder, includeDefaults, params);
+        if (nullValue != null) {
+            builder.field("null_value", nullValue);
+        }
+    }
+
+    @Override
+    public Iterator<Mapper> iterator() {
+        List<Mapper> subIterators = new ArrayList<>();
+        subIterators.add(prereleaseSubField);
+        subIterators.add(majorVersionSubField);
+        subIterators.add(minorVersionSubField);
+        subIterators.add(patchVersionSubField);
+        @SuppressWarnings("unchecked")
+        Iterator<Mapper> concat = Iterators.concat(super.iterator(), subIterators.iterator());
+        return concat;
+    }
+
+    @Override
+    protected Object parseSourceValue(Object value, String format) {
+        if (format != null) {
+            throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] doesn't support formats.");
+        }
+        return value.toString();
+    }
+
+    private static DocValueFormat VERSION_DOCVALUE = new DocValueFormat() {
```

Comment:
Do you need to register this format in `NamedWriteableRegistry` like we register other format in `SearchModule#registerValueFormats`?

Reply:
Good point, I guess this hasn't shown up in tests yet because they are mostly single node. I'll see if this would e.g. fail a yaml aggs test. I can probably register this in the plugin.

Reply:
I added c70e7c1 to register the format and an IT test that would have caught this when serializing aggregations across nodes. Thanks for catching this!